File: app/src/main/java/net/north101/android/ghplayertracker/BaseViewHolder.java
package net.north101.android.ghplayertracker;

import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

public abstract class BaseViewHolder<T> extends RecyclerView.ViewHolder implements View.OnClickListener, View.OnLongClickListener {
    public T item;
    private ClickListener<T> listener;

    public static View inflate(ViewGroup parent, int layout) {
        return LayoutInflater.from(parent.getContext()).inflate(layout, parent, false);
    }

    public BaseViewHolder(View itemView) {
        super(itemView);
        this.itemView.setOnClickListener(this);
        this.itemView.setOnLongClickListener(this);
    }

    public void bind(T item) {
        this.item = item;

    }

    public void unbind() {
        this.item = null;
        this.listener = null;
    }

    public void setOnItemClickListener(ClickListener<T> listener) {
        this.listener = listener;
    }

    @Override
    public void onClick(View view) {
        if (this.listener == null) return;
        this.listener.onItemClick(this);
    }

    @Override
    public boolean onLongClick(View view) {
        if (this.listener == null) return false;
        Log.d("test", "longclick");
        return this.listener.onItemLongClick(this);
    }

    public static class ClickListener<T> {
        public void onItemClick(BaseViewHolder<T> holder) {}
        public boolean onItemLongClick(BaseViewHolder<T> holder) {
            return false;
        }
    }
}

File: app/src/main/java/net/north101/android/ghplayertracker/BaseViewHolder.kt
package net.north101.android.ghplayertracker

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

abstract class BaseViewHolder<T>(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener, View.OnLongClickListener {
    var item: T? = null
    private var listener: ClickListener<T>? = null

    init {
        this.itemView.setOnClickListener(this)
        this.itemView.setOnLongClickListener(this)
    }

    open fun bind(item: T) {
        this.item = item

    }

    fun unbind() {
        this.item = null
        this.listener = null
    }

    fun setOnItemClickListener(listener: ClickListener<T>) {
        this.listener = listener
    }

    override fun onClick(view: View) {
        if (this.listener == null) return
        this.listener!!.onItemClick(this)
    }

    override fun onLongClick(view: View): Boolean {
        if (this.listener == null) return false
        return this.listener!!.onItemLongClick(this)
    }

    open class ClickListener<T> {
        open fun onItemClick(holder: BaseViewHolder<T>) {}
        open fun onItemLongClick(holder: BaseViewHolder<T>): Boolean {
            return false
        }
    }

    companion object {

        fun inflate(parent: ViewGroup, layout: Int): View {
            return LayoutInflater.from(parent.context).inflate(layout, parent, false)
        }
    }
}
